### PR TITLE
Rule evaluation:  Allow Bundles with PATCH Parameters

### PR DESCRIPTION
**Pull request by lukedegruchy**
Closes https://github.com/hapifhir/hapi-fhir/issues/5642

### CODE REVIEW

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleImplOpTest.java [hunk 1]
```diff
@@ -93,4 +99,14 @@ public void testMatchesAllResources() {
...
 	}
+
+	@Test
+	void testTransactionBundleWithParameters() {
+		final RuleImplOp ruleOp = new RuleImplOp("transactionBundlePatch");
+		ruleOp.setOp(RuleOpEnum.TRANSACTION);
+
+		final IBaseResource mockResouce = Mockito.mock(IBaseResource.class);
+
+//		ruleOp.applyRule(RestOperationTypeEnum.TRANSACTION, new SystemRequestDetails(), mockResouce, null, null, null, null);
```
Comment (michaelabuckley) [inline hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleImplOpTest.java:110]: problem: looks like the test is commented out.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/FhirQueryRuleImplTest.java [hunk 2]
```diff
@@ -251,8 +273,99 @@ public void allowIdwithQueryFilter_matchesJustIdNotFilter_abstain() {
...
+
+		@Test
+		void testTransactionBundleWithNestedBundle() {
+			final Parameters parameters = buildParameters();
+
+			final IBaseBundle bundle = getBundle(true, parameters);
+
+			// There's no better way to "put a Bundle inside a Bundle" with the BundleBuilder
+			final RuntimeResourceDefinition mockRuntimeResourceDefinition  = mock(RuntimeResourceDefinition.class);
+			when(mockRuntimeResourceDefinition.getName()).thenReturn(BUNDLE);
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/FhirQueryRuleImplTest.java:314]: problem:  I'm very skeptical we ever need to mock the FhirContext.  This test uses a concrete one.

**Review by michaelabuckley — CHANGES_REQUESTED** [review]
Please re-write these tests without mock.  You should be able to use the normal RuleBuilder and make normal assertions.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/FhirQueryRuleImplTest.java (hunk 2 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/FhirQueryRuleImplTest.java:314]: I need to add a Bundle to a Bundle, which the BundleBuilder doesn't allow me to do.  If it does allow me to, it's not at all obvious.

File: hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleImplOpTest.java (hunk 1 above)
Comment (lukedegruchy) [reply hapi-fhir-server/src/test/java/ca/uhn/fhir/rest/server/interceptor/auth/RuleImplOpTest.java]: I forgot to revert my changes to this test after I realized I didn't have access to Parameters.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/RuleImplOpTest.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+import java.util.HashSet;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.fail;
+import static org.mockito.Mockito.when;
+
+@MockitoSettings
+public class RuleImplOpTest {
```
Comment (michaelabuckley) [inline hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/RuleImplOpTest.java:31]: This looks nice and concrete.  Thanks.

question: does this test fail without your fix?

**Review by michaelabuckley — APPROVED** [review]
looks good.

File: hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/RuleImplOpTest.java (hunk 3 above)
Comment (lukedegruchy) [reply hapi-fhir-jpaserver-test-utilities/src/test/java/ca/uhn/fhir/jpa/auth/RuleImplOpTest.java:31]: `testTransactionBundlePatchWithParameters()` fails without my fix.  The other two tests pass, as I expect them to

**Pull request merged**